File: evals/harness_compare/lib/adapters/jewell.py
from __future__ import annotations

import json
import os
import subprocess
import time
import urllib.error
import urllib.request
from pathlib import Path

from ..task import Task
from .base import AdapterResult, BaseAdapter

# Monorepo root: evals/harness_compare/lib/adapters/this → parents[4]
REPO_ROOT = Path(__file__).resolve().parents[4]
RUST_AGENT = REPO_ROOT / "rust-agent"


class JewellAdapter(BaseAdapter):
    """Drive jewell-labs rust-agent (host evals + chat stream)."""

    name = "jewell"
    provides = {
        "host_eval",
        "agent_os",
        "closed_form",
        "multi_step",
        # coding/terminal intentionally NOT claimed — honest capability surface
    }

    def __init__(self) -> None:
        self.base = os.environ.get("JEWELL_BASE", "http://127.0.0.1:3000").rstrip("/")
        # API expects category/name (see load_agent), not bare stem.
        self.agent = os.environ.get("JEWELL_AGENT", "core/orchestrator")
        self.llama = os.environ.get("COMPARE_LLAMA", "http://127.0.0.1:8080").rstrip("/")

    def run(self, task: Task, workspace: Path, instruction: str, timeout_s: int) -> AdapterResult:
        ok, missing = self.can_run(task)
        if not ok:
            return AdapterResult(
                status="skip",
                detail=f"jewell lacks capabilities: {missing}",
                capabilities_missing=missing,
            )

        if "host_eval" in task.capabilities or task.track == "agent_os" and task.id.startswith(
            "host_"
        ):
            return self._run_host_eval(task, timeout_s)

        if task.track == "closed_form" and os.environ.get("JEWELL_CLOSED_FORM", "chat") == "direct":
            return self._run_llama_direct(instruction, timeout_s)

        return self._run_chat(task, workspace, instruction, timeout_s)

    def _run_host_eval(self, task: Task, timeout_s: int) -> AdapterResult:
        """Structural gates without requiring a live chat model."""
        mode = (task.path / "workspace" / "host_mode.txt").read_text(encoding="utf-8").strip() if (
            task.path / "workspace" / "host_mode.txt"
        ).is_file() else "tool_plan"

        t0 = time.perf_counter()
        if mode == "tool_plan":
            cmd = ["cargo", "run", "-q", "--bin", "eval_introspection"]
            try:
                proc = subprocess.run(
                    cmd,
                    cwd=str(RUST_AGENT),
                    capture_output=True,
                    text=True,
                    timeout=timeout_s,
                )
            except (subprocess.TimeoutExpired, OSError) as e:
                return AdapterResult(status="error", detail=f"host eval: {e}")
            t_ms = int((time.perf_counter() - t0) * 1000)
            # Write a marker file the grader can check; also return stdout for debug.
            return AdapterResult(
                status="ok" if proc.returncode == 0 else "error",
                stdout=proc.stdout or "",
                stderr=proc.stderr or "",
                t_ms=t_ms,
                detail="eval_introspection",
                extra={"exit_code": proc.returncode, "host_mode": mode},
            )

        if mode == "team":
            # Re-use unit test as host gate
            cmd = [
                "cargo",
                "test",
                "-q",
                "team_collaboration_green",
                "--",
                "--nocapture",
            ]
            try:
                proc = subprocess.run(
                    cmd,
                    cwd=str(RUST_AGENT),
                    capture_output=True,
                    text=True,
                    timeout=timeout_s,
                )
            except (subprocess.TimeoutExpired, OSError) as e:
                return AdapterResult(status="error", detail=f"team host eval: {e}")
            t_ms = int((time.perf_counter() - t0) * 1000)
            return AdapterResult(
                status="ok" if proc.returncode == 0 else "error",
                stdout=proc.stdout or "",
                stderr=proc.stderr or "",
                t_ms=t_ms,
                detail="team_collaboration_green",
                extra={"exit_code": proc.returncode, "host_mode": mode},
            )

        return AdapterResult(status="error", detail=f"unknown host_mode {mode}")

    def _run_chat(
        self, task: Task, workspace: Path, instruction: str, timeout_s: int
    ) -> AdapterResult:
        prompt = (
            f"Evaluation workspace directory: {workspace}\n"
            f"Task track: {task.track}\n"
            f"Use your available tools. Prefer concise final answers.\n"
            f"If the task asks for files, explain what you would write "
            f"(you may lack arbitrary FS write tools).\n\n"
            f"{instruction.strip()}\n"
        )
        body = json.dumps(
            {
                "agent_stem": self.agent,
                "message": prompt,
                "history": [],
            }
        ).encode("utf-8")
        req = urllib.request.Request(
            f"{self.base}/api/chat/stream",
            data=body,
            headers={"Content-Type": "application/json", "Accept": "text/event-stream"},
            method="POST",
        )
        t0 = time.perf_counter()
        chunks: list[str] = []
        stream_err = ""
        try:
            with urllib.request.urlopen(req, timeout=timeout_s) as resp:
                # SSE: data: {...}\n\n  (token deltas use {"delta": "..."})
                buf = ""
                done = False
                while not done:
                    piece = resp.read(4096)
                    if not piece:
                        break
                    buf += piece.decode("utf-8", errors="replace")
                    while "\n" in buf:
                        line, buf = buf.split("\n", 1)
                        line = line.strip()
                        if not line.startswith("data:"):
                            continue
                        data = line[5:].strip()
                        if data == "[DONE]":
                            done = True
                            break
                        try:
                            obj = json.loads(data)
                        except json.JSONDecodeError:
                            chunks.append(data)
                            continue
                        if "error" in obj:
                            # Keep partial transcript — context overflow mid-run is common
                            # on small llama-server ctx; graders may still pass.
                            stream_err = str(obj["error"])
                            done = True
                            break
                        for k in ("text", "content", "delta", "message", "token"):
                            if k in obj and isinstance(obj[k], str):
                                chunks.append(obj[k])
                        if obj.get("type") == "assistant" and isinstance(obj.get("content"), str):
                            chunks.append(obj["content"])
        except urllib.error.HTTPError as e:
            t_ms = int((time.perf_counter() - t0) * 1000)
            err = e.read().decode("utf-8", errors="replace") if e.fp else str(e)
            return AdapterResult(
                status="error",
                stderr=err,
                t_ms=t_ms,
                detail=f"HTTP {e.code} from {self.base} — is rust-agent running?",
            )
        except urllib.error.URLError as e:
            return AdapterResult(
                status="error",
                detail=f"cannot reach jewell at {self.base}: {e.reason}",
            )
        except TimeoutError:
            return AdapterResult(status="error", detail="chat timeout")

        t_ms = int((time.perf_counter() - t0) * 1000)
        text = "".join(chunks)
        if not text.strip():
            return AdapterResult(
                status="error",
                stderr=stream_err,
                t_ms=t_ms,
                detail="chat error (empty transcript)",
            )

        if task.track == "closed_form":
            (workspace / "agent_answer.txt").write_text(text, encoding="utf-8")
        else:
            (workspace / "agent_transcript.txt").write_text(text, encoding="utf-8")

        detail = "jewell chat stream completed"
        if stream_err:
            detail = f"partial stream (server error after tokens): {stream_err[:120]}"
        return AdapterResult(
            status="ok",
            stdout=text,
            stderr=stream_err,
            t_ms=t_ms,
            detail=detail,
        )

    def _run_llama_direct(self, instruction: str, timeout_s: int) -> AdapterResult:
        """Optional closed_form path: raw llama-server (no agent tools)."""
        body = json.dumps(
            {
                "model": "local",
                "temperature": 0,
                "messages": [{"role": "user", "content": instruction}],
            }
        ).encode("utf-8")
        req = urllib.request.Request(
            f"{self.llama}/v1/chat/completions",
            data=body,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        t0 = time.perf_counter()
        try:
            with urllib.request.urlopen(req, timeout=timeout_s) as resp:
                payload = json.loads(resp.read().decode("utf-8"))
        except Exception as e:
            return AdapterResult(status="error", detail=f"llama direct: {e}")
        t_ms = int((time.perf_counter() - t0) * 1000)
        try:
            text = payload["choices"][0]["message"]["content"]
        except (KeyError, IndexError, TypeError):
            return AdapterResult(
                status="error",
                stdout=json.dumps(payload)[:2000],
                t_ms=t_ms,
                detail="bad completion shape",
            )
        return AdapterResult(status="ok", stdout=text, t_ms=t_ms, detail="llama direct")

File: rust-agent/src/eval/compare/adapters/jewell.rs
//! Jewell harness: in-process host gates + optional HTTP chat for closed_form.

use super::{AdapterResult, HarnessAdapter};
use crate::eval::compare::task::Task;
use crate::eval::{
    orchestrator_ctx, run_team_collaboration_case, run_tool_plan_case, GroundTruth,
};
use std::fs;
use std::path::Path;
use std::time::Instant;

pub struct JewellAdapter {
    base: String,
    agent: String,
}

impl JewellAdapter {
    pub fn new() -> Self {
        Self {
            base: std::env::var("JEWELL_BASE").unwrap_or_else(|_| "http://127.0.0.1:3000".into()),
            agent: std::env::var("JEWELL_AGENT")
                .unwrap_or_else(|_| "core/orchestrator".into()),
        }
    }
}

impl HarnessAdapter for JewellAdapter {
    fn name(&self) -> &str {
        "jewell"
    }

    fn provides(&self) -> &[&str] {
        // Product chat + host gates + agent sandbox fs_* (write_file surface).
        &[
            "host_eval",
            "agent_os",
            "closed_form",
            "multi_step",
            "write_file",
            "read_file",
            "terminal",
            "coding",
        ]
    }

    fn run(&self, task: &Task, workspace: &Path, instruction: &str) -> AdapterResult {
        let (ok, missing) = self.can_run(task);
        if !ok {
            return AdapterResult::skip(
                format!("jewell lacks capabilities: {missing:?}"),
                missing,
            );
        }

        if task.capabilities.iter().any(|c| c == "host_eval") {
            return self.run_host_eval(task, workspace);
        }

        self.run_chat(task, workspace, instruction)
    }
}

impl JewellAdapter {
    fn run_host_eval(&self, task: &Task, workspace: &Path) -> AdapterResult {
        let mode_path = task.workspace_src().join("host_mode.txt");
        let mode = fs::read_to_string(&mode_path)
            .unwrap_or_else(|_| "tool_plan".into())
            .trim()
            .to_string();

        let t0 = Instant::now();
        let (exit, stdout, detail) = match mode.as_str() {
            "team" => {
                let case = run_team_collaboration_case();
                let code = if case.correct { 0 } else { 1 };
                (
                    code,
                    format!("team_collaboration correct={}", case.correct),
                    "team_collaboration_green".to_string(),
                )
            }
            _ => {
                // In-process tool_plan + team (no cargo subprocess)
                let (ctx, _) = orchestrator_ctx();
                let gt = GroundTruth::collect(&ctx);
                let plan = run_tool_plan_case(&ctx, &gt);
                let team = run_team_collaboration_case();
                let code = if plan.correct && team.correct { 0 } else { 1 };
                (
                    code,
                    format!(
                        "tool_plan={} team={}",
                        plan.correct, team.correct
                    ),
                    "eval_introspection".to_string(),
                )
            }
        };
        let _ = fs::write(workspace.join("host_exit_code.txt"), format!("{exit}\n"));
        AdapterResult {
            status: if exit == 0 { "ok" } else { "error" }.into(),
            stdout,
            stderr: String::new(),
            t_ms: t0.elapsed().as_millis() as u64,
            detail,
            capabilities_missing: vec![],
            exit_code: Some(exit),
        }
    }

    fn run_chat(&self, task: &Task, workspace: &Path, instruction: &str) -> AdapterResult {
        // Fairness: shared instruction only — no coach wrappers.
        let _ = (task, workspace);
        let body = serde_json::json!({
            "agent_stem": self.agent,
            "message": instruction.trim(),
            "history": [],
        });
        let url = format!("{}/api/chat/stream", self.base.trim_end_matches('/'));
        let t0 = Instant::now();
        let client = match reqwest::blocking::Client::builder()
            .timeout(std::time::Duration::from_secs(task.timeout_s.max(30)))
            .build()
        {
            Ok(c) => c,
            Err(e) => return AdapterResult::error(format!("http client: {e}")),
        };
        let resp = match client
            .post(&url)
            .header("content-type", "application/json")
            .header("accept", "text/event-stream")
            .json(&body)
            .send()
        {
            Ok(r) => r,
            Err(e) => {
                return AdapterResult::error(format!(
                    "cannot reach jewell at {}: {e}",
                    self.base
                ));
            }
        };
        if !resp.status().is_success() {
            let code = resp.status().as_u16();
            let err = resp.text().unwrap_or_default();
            return AdapterResult {
                status: "error".into(),
                stdout: String::new(),
                stderr: err,
                t_ms: t0.elapsed().as_millis() as u64,
                detail: format!("HTTP {code} from {} — is rust-agent running?", self.base),
                capabilities_missing: vec![],
                exit_code: None,
            };
        }
        let raw = resp.text().unwrap_or_default();
        let mut chunks = Vec::new();
        let mut stream_err = String::new();
        for line in raw.lines() {
            let line = line.trim();
            let Some(data) = line.strip_prefix("data:") else {
                continue;
            };
            let data = data.trim();
            if data.is_empty() || data == "[DONE]" {
                continue;
            }
            let Ok(obj) = serde_json::from_str::<serde_json::Value>(data) else {
                chunks.push(data.to_string());
                continue;
            };
            if let Some(e) = obj.get("error") {
                stream_err = e.to_string();
                break;
            }
            for k in ["text", "content", "delta", "message", "token"] {
                if let Some(s) = obj.get(k).and_then(|x| x.as_str()) {
                    chunks.push(s.to_string());
                }
            }
        }
        let text = chunks.join("");
        if text.trim().is_empty() {
            return AdapterResult {
                status: "error".into(),
                stdout: String::new(),
                stderr: stream_err,
                t_ms: t0.elapsed().as_millis() as u64,
                detail: "chat error (empty transcript)".into(),
                capabilities_missing: vec![],
                exit_code: None,
            };
        }
        if task.track == "closed_form" {
            let _ = fs::write(workspace.join("agent_answer.txt"), &text);
        } else {
            let _ = fs::write(workspace.join("agent_transcript.txt"), &text);
        }
        let detail = if stream_err.is_empty() {
            "jewell chat stream completed".into()
        } else {
            format!("partial stream: {}", &stream_err[..stream_err.len().min(120)])
        };
        AdapterResult {
            status: "ok".into(),
            stdout: text,
            stderr: stream_err,
            t_ms: t0.elapsed().as_millis() as u64,
            detail,
            capabilities_missing: vec![],
            exit_code: Some(0),
        }
    }
}
